### PR TITLE
[mlir][LLVM] Add disjoint flag

### PR DESCRIPTION
**Pull request by lfrenot**
The implementation is mostly based on the one existing for the exact flag.

disjoint means that for each bit, that bit is zero in at least one of the inputs. This allows the Or to be treated as an Add since no carry can occur from any bit. If the disjoint keyword is present, the result value of the or is a [poison value](https://llvm.org/docs/LangRef.html#poisonvalues) if both inputs have a one in the same bit position. For vectors, only the element containing the bit is poison.

### CODE REVIEW

**Comment by lfrenot** [comment]
@zero9178, @gysit and @Dinistro, could you take a look?

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:114]:
```suggestion
  string llvmBuilder = [{
    auto inst = builder.Create}] # instName # [{($lhs, $rhs, /*Name=*/"");
    moduleTranslation.setDisjointFlag(op, inst);
    $res = inst;
  }];
```
nit: an attempt to format tablegen a bit nicer, feel free to pick it if it works or ignore it otherwise.

**Review by gysit — APPROVED** [review]
Amazing how many flags LLVM has :).

Thanks LGTM modulo nits.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMInterfaces.td
Comment (lfrenot) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMInterfaces.td]: Done.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp]: Done.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Done.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (lfrenot) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td]: It seems to work, done.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1903]:
```suggestion
  auto disjointInst = cast<llvm::PossiblyDisjointInst>(inst);
```
nit: missed this before.

**Review by gysit — COMMENTED** [review]
Thanks for addressing the comment! Just saw two small things while looking at your changes. Once fixed, I will wait for the build bots and land.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Done.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Done.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Can you rename inst_disjoint in the other function as well? Sorry I forgot to mark that as well. Feel free to fix this things proactively.

Regarding the llvm::Value. I think you could use llvm::Instruction and it would cast to a llvm::Value right? That maybe preferable over using a value since it is in line with all the other helper functions we have. If it doesn't work the comment on the function needs to be fixed by replacing inst -> value.

Sorry for being a bit pedantic :).

Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Sorry, I forgot about the other function.

As for the Value issue, I may have been unclear, but the issue I get if I use Instruction is from the llvm builder
```
/home/leon_frenot/fork-llvm/build/tools/mlir/include/mlir/Dialect/LLVMIR/LLVMConversions.inc:279:43: error: invalid conversion from ‘llvm::Value*’ to ‘llvm::Instruction*’ [-fpermissive]
  279 |     moduleTranslation.setDisjointFlag(op, inst);
         |                                                                 ^~~~
         |                                                                 |
         |                                                                 llvm::Value*
```
So the issue comes from builder.CreateOr returning a Value instead of an Instruction.

For now I'll simply fix the comment

Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Done.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: That makes sense!

**Comment by gysit** [comment]
It looks like there is actually some issue with a flang test (see red CI runs). I am not sure if this is new or if this has been there before the last nit changes.

**Comment by lfrenot** [comment]
Apparently, it's been here since the first commit: https://buildkite.com/llvm-project/github-pull-requests/builds/118691
I'll look into it, there seems to be an issue with the case in moduleTranslate

**Comment by lfrenot** [comment]
This should fix the flang tests @gysit 
But I'm not sure it is a proper way to do it

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1904]: What operation produces the error on flang? I would expect this function is only called for llvm.or operation? Shouldn't the generated operation also implement the interface? Or may the llvm instruction fold somehow / be a constant?

I am not agains the solution but it would be nice to understand what the root cause is and then add a comment why the cast may fail.

Comment (lfrenot) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1904]: I've looked into it, and yes, what happens is that the instruction gets folded into a constant by CreateOr.

Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1905]: Thanks for investigating.

**Pull request merged**
